apps/homescreen/js/app.js

```diff
@@ -540,6 +577,12 @@ const SETTINGS_VERSION = 0;
 
         this.pageHeight = pageHeight;
         this.pendingGridHeight = gridHeight;
+
+        if (!this.visualLoadComplete &&
+            Math.floor(visibleChildren / (this.small ? 4 : 3)) * iconHeight >=
+              scrollHeight) {
+          this.onVisualLoad();
+        }
```

Comment:
I don't entirely agree here. Only the icons on the first visible page are populated, but the icons on the first row of the next page are usually visible (on a Z3C screen aspect ratio at least). We should really call `onVisualLoad()` when all the icons above the fold are visible.
Have you tried `Math.floor(visibleChildren / (this.small ? 5 : 4))`? If so does it result in significantly worse result?